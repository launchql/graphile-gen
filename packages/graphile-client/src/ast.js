import * as t from '@pyramation/graphql-ast';
import plz from 'pluralize';
import inflection from 'inflection';
const NON_MUTABLE_PROPS = [
  'id',
  'createdAt',
  'createdBy',
  'updatedAt',
  'updatedBy'
];
const objectToArray = (obj) =>
  Object.keys(obj).map((k) => ({ name: k, ...obj[k] }));

const createGqlMutation = ({
  operationName,
  mutationName,
  selectArgs,
  selections,
  variableDefinitions,
  modelName,
  useModel = true
}) => {
  const opSel = !modelName
    ? [
        t.field({
          name: operationName,
          args: selectArgs,
          selectionSet: t.selectionSet({ selections })
        })
      ]
    : [
        t.field({
          name: operationName,
          args: selectArgs,
          selectionSet: t.selectionSet({
            selections: useModel
              ? [
                  t.field({
                    name: modelName,
                    selectionSet: t.selectionSet({ selections })
                  })
                ]
              : selections
          })
        })
      ];

  return t.document({
    definitions: [
      t.operationDefinition({
        operation: 'mutation',
        name: mutationName,
        variableDefinitions,
        selectionSet: t.selectionSet({ selections: opSel })
      })
    ]
  });
};

export const getAll = ({ queryName, operationName, query, fields }) => {
  const selections = getSelections(query, fields);

  const opSel = [
    t.field({
      name: operationName,
      selectionSet: t.objectValue({
        fields: [
          t.field({
            name: 'totalCount'
          }),
          t.field({
            name: 'nodes',
            selectionSet: t.selectionSet({ selections })
          })
        ]
      })
    })
  ];

  const ast = t.document({
    definitions: [
      t.operationDefinition({
        operation: 'query',
        name: queryName,
        selectionSet: t.selectionSet({ selections: opSel })
      })
    ]
  });

  return ast;
};

export const getMany = ({
  client, // we can use props here to enable pagination, etc
  queryName,
  operationName,
  query,
  fields
}) => {
  const Singular = query.model;
  const Plural = operationName.charAt(0).toUpperCase() + operationName.slice(1);
  const Condition = `${Singular}Condition`;
  const Filter = `${Singular}Filter`;
  const OrderBy = `${Plural}OrderBy`;
  const selections = getSelections(query, fields);

  const ast = t.document({
    definitions: [
      t.operationDefinition({
        operation: 'query',
        name: queryName,
        variableDefinitions: [
          t.variableDefinition({
            variable: t.variable({
              name: 'first'
            }),
            type: t.namedType({
              type: 'Int'
            })
          }),
          t.variableDefinition({
            variable: t.variable({
              name: 'last'
            }),
            type: t.namedType({
              type: 'Int'
            })
          }),
          t.variableDefinition({
            variable: t.variable({
              name: 'after'
            }),
            type: t.namedType({
              type: 'Cursor'
            })
          }),
          t.variableDefinition({
            variable: t.variable({
              name: 'before'
            }),
            type: t.namedType({
              type: 'Cursor'
            })
          }),
          t.variableDefinition({
            variable: t.variable({
              name: 'offset'
            }),
            type: t.namedType({
              type: 'Int'
            })
          }),
          t.variableDefinition({
            variable: t.variable({
              name: 'condition'
            }),
            type: t.namedType({
              type: Condition
            })
          }),
          t.variableDefinition({
            variable: t.variable({
              name: 'filter'
            }),
            type: t.namedType({
              type: Filter
            })
          }),
          t.variableDefinition({
            variable: t.variable({
              name: 'orderBy'
            }),
            type: t.listType({
              type: t.nonNullType({ type: t.namedType({ type: OrderBy }) })
            })
          })
        ],
        selectionSet: t.selectionSet({
          selections: [
            t.field({
              name: operationName,
              args: [
                t.argument({
                  name: 'first',
                  value: t.variable({
                    name: 'first'
                  })
                }),
                t.argument({
                  name: 'last',
                  value: t.variable({
                    name: 'last'
                  })
                }),
                t.argument({
                  name: 'offset',
                  value: t.variable({
                    name: 'offset'
                  })
                }),
                t.argument({
                  name: 'after',
                  value: t.variable({
                    name: 'after'
                  })
                }),
                t.argument({
                  name: 'before',
                  value: t.variable({
                    name: 'before'
                  })
                }),
                t.argument({
                  name: 'condition',
                  value: t.variable({
                    name: 'condition'
                  })
                }),
                t.argument({
                  name: 'filter',
                  value: t.variable({
                    name: 'filter'
                  })
                }),
                t.argument({
                  name: 'orderBy',
                  value: t.variable({
                    name: 'orderBy'
                  })
                })
              ],
              selectionSet: t.objectValue({
                fields: [
                  t.field({
                    name: 'totalCount'
                  }),
                  t.field({
                    name: 'pageInfo',
                    selectionSet: t.selectionSet({
                      selections: [
                        t.field({ name: 'hasNextPage' }),
                        t.field({ name: 'hasPreviousPage' }),
                        t.field({ name: 'endCursor' }),
                        t.field({ name: 'startCursor' })
                      ]
                    })
                  }),
                  client._edges
                    ? t.field({
                        name: 'edges',
                        selectionSet: t.selectionSet({
                          selections: [
                            t.field({ name: 'cursor' }),
                            t.field({
                              name: 'node',
                              selectionSet: t.selectionSet({ selections })
                            })
                          ]
                        })
                      })
                    : t.field({
                        name: 'nodes',
                        selectionSet: t.selectionSet({
                          selections
                        })
                      })
                ]
              })
            })
          ]
        })
      })
    ]
  });

  return ast;
};

export const getOne = ({
  client, // we can use props here to enable pagination, etc
  queryName,
  operationName,
  query,
  fields
}) => {
  const variableDefinitions = Object.keys(query.properties)
    .map((key) => ({ name: key, ...query.properties[key] }))
    .filter((field) => field.isNotNull)
    .map((field) => {
      const {
        name: fieldName,
        type: fieldType,
        isNotNull,
        isArray,
        isArrayNotNull
      } = field;
      let type = t.namedType({ type: fieldType });
      if (isNotNull) type = t.nonNullType({ type });
      if (isArray) {
        type = t.listType({ type });
        if (isArrayNotNull) type = t.nonNullType({ type });
      }
      return t.variableDefinition({
        variable: t.variable({ name: fieldName }),
        type
      });
    });

  const props = objectToArray(query.properties);

  const selectArgs = props
    .filter((field) => field.isNotNull)
    .map((field) => {
      return t.argument({
        name: field.name,
        value: t.variable({ name: field.name })
      });
    });

  const selections = getSelections(query, fields);
  const opSel = [
    t.field({
      name: operationName,
      args: selectArgs,
      selectionSet: t.selectionSet({ selections })
    })
  ];

  const ast = t.document({
    definitions: [
      t.operationDefinition({
        operation: 'query',
        name: queryName,
        variableDefinitions,
        selectionSet: t.selectionSet({ selections: opSel })
      })
    ]
  });
  return ast;
};

export const createOne = ({ mutationName, operationName, mutation }) => {
  if (!mutation.properties?.input?.properties) {
    console.log('no input field for mutation for' + mutationName);
    return;
  }

  const modelName = inflection.camelize(
    [plz.singular(mutation.model)].join('_'),
    true
  );

  const allAttrs = objectToArray(
    mutation.properties.input.properties[modelName].properties
  );

  const attrs = allAttrs.filter(
    (field) => !NON_MUTABLE_PROPS.includes(field.name)
  );

  const variableDefinitions = attrs.map((field) => {
    const {
      name: fieldName,
      type: fieldType,
      isNotNull,
      isArray,
      isArrayNotNull
    } = field;
    let type = t.namedType({ type: fieldType });
    if (isNotNull) type = t.nonNullType({ type });
    if (isArray) {
      type = t.listType({ type });
      if (isArrayNotNull) type = t.nonNullType({ type });
    }
    return t.variableDefinition({
      variable: t.variable({ name: fieldName }),
      type
    });
  });

  const selectArgs = [
    t.argument({
      name: 'input',
      value: t.objectValue({
        fields: [
          t.objectField({
            name: modelName,
            value: t.objectValue({
              fields: attrs.map((field) =>
                t.objectField({
                  name: field.name,
                  value: t.variable({
                    name: field.name
                  })
                })
              )
            })
          })
        ]
      })
    })
  ];

  const selections = allAttrs.map((field) => t.field({ name: field.name }));
  const ast = createGqlMutation({
    operationName,
    mutationName,
    selectArgs,
    selections,
    variableDefinitions,
    modelName
  });

  return ast;
};

export function getSelections(query, fields = []) {
  const useAll = fields.length === 0;

  return query.selection
    .map((field) => {
      if (!useAll && !fields.includes(field)) return null;
      if (typeof field === 'object' && field !== null) {
        return t.field({
          name: field.name,
          args: [
            t.argument({
              name: 'first',
              value: t.intValue({ value: 3 })
            })
          ],
          selectionSet: t.objectValue({
            fields: [
              t.field({
                name: 'nodes',
                selectionSet: t.selectionSet({
                  selections: field.selection.map((field) =>
                    t.field({ name: field })
                  )
                })
              })
            ]
          })
        });
      }
      return t.field({ name: field });
    })
    .filter((i) => Boolean(i));
}
