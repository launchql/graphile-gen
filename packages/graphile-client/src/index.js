export default class GraphileClient {
  constructor (meta) {
    this.meta = meta;
  }
}