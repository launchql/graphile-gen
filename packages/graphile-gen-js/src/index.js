export default {
  yo: 1,
};
