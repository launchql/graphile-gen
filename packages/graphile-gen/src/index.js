import plz from 'pluralize';
import inflection from 'inflection';
import * as t from '@pyramation/graphql-ast';

function getType(type) {
  // TODO look in postgraphile
  switch (type) {
    case 'uuid':
      return 'UUID';
    case 'json':
    case 'jsonb':
      return 'JSON';
    case 'numeric':
      return 'BigFloat';
    case 'int':
    case 'integer':
      return 'Int';
    case 'upload':
    case 'attachment':
    case 'image':
      return 'Upload';
    case 'text':
    default:
      return 'String';
  }
}

export const createOne = (defn) => {
  const operationName = inflection.camelize(
    ['create', plz.singular(defn.name)].join('_'),
    true
  );
  const mutationName = inflection.camelize(
    ['create', plz.singular(defn.name), 'mutation'].join('_'),
    true
  );
  const modelName = inflection.camelize(
    [plz.singular(defn.name)].join('_'),
    true
  );

  const fieldMap = defn.fields.nodes.reduce((m, field) => {
    const fieldName = inflection.camelize(plz.singular(field.name), true);
    const fieldType = getType(field.type);
    const isArray = !!field.isArray;
    const isRequired =
      field.isRequired &&
      (typeof field.defaultValue === 'undefined' ||
        field.defaultValue === null);
    m[field.name] = {
      fieldName,
      fieldType,
      isRequired,
      isArray
    };
    return m;
  }, {});

  const variableDefinitions = defn.fields.nodes.map((field) => {
    const { fieldName, fieldType, isRequired, isArray } = fieldMap[field.name];
    let type = t.namedType({ type: fieldType });
    if (isArray) type = t.listType({ type });
    if (isRequired) type = t.nonNullType({ type });
    return t.variableDefinition({
      variable: t.variable({ name: fieldName }),
      type
    });
  });

  const selectArgs = [
    t.argument({
      name: 'input',
      value: t.objectValue({
        fields: [
          t.objectField({
            name: modelName,
            value: t.objectValue({
              fields: defn.fields.nodes.map((field) =>
                t.objectField({
                  name: fieldMap[field.name].fieldName,
                  value: t.variable({
                    name: fieldMap[field.name].fieldName
                  })
                })
              )
            })
          })
        ]
      })
    })
  ];

  const selections = defn.fields.nodes.map((field) =>
    t.field({ name: fieldMap[field.name].fieldName })
  );
  const opSel = [
    t.field({
      name: operationName,
      args: selectArgs,
      selectionSet: t.selectionSet({
        selections: [
          t.field({
            name: modelName,
            selectionSet: t.selectionSet({ selections })
          })
        ]
      })
    })
  ];

  const ast = t.document({
    definitions: [
      t.operationDefinition({
        operation: 'mutation',
        name: mutationName,
        variableDefinitions,
        selectionSet: t.selectionSet({ selections: opSel })
      })
    ]
  });

  return { name: mutationName, ast };
};

// TODO use constraints...

const PATCHER_FIELD = 'id';
export const updateOne = (defn) => {
  const operationName = inflection.camelize(
    ['update', plz.singular(defn.name)].join('_'),
    true
  );
  const mutationName = inflection.camelize(
    ['update', plz.singular(defn.name), 'mutation'].join('_'),
    true
  );
  const modelName = inflection.camelize(
    [plz.singular(defn.name)].join('_'),
    true
  );

  const fieldMap = defn.fields.nodes.reduce((m, field) => {
    const fieldName = inflection.camelize(plz.singular(field.name), true);
    const fieldType = getType(field.type);
    const isArray = !!field.isArray;
    const isRequired =
      field.isRequired &&
      (typeof field.defaultValue === 'undefined' ||
        field.defaultValue === null);
    m[field.name] = {
      fieldName,
      fieldType,
      isRequired,
      isArray
    };
    return m;
  }, {});

  const variableDefinitions = defn.fields.nodes.map((field) => {
    const { fieldName, fieldType, isRequired, isArray } = fieldMap[field.name];
    let type = t.namedType({ type: fieldType });
    if (isArray) type = t.listType({ type });
    // later we can use primary key info...
    if (field.name === PATCHER_FIELD) type = t.nonNullType({ type });
    return t.variableDefinition({
      variable: t.variable({ name: fieldName }),
      type
    });
  });

  const selectArgs = [
    t.argument({
      name: 'input',
      value: t.objectValue({
        fields: [
          t.objectField({
            name: PATCHER_FIELD,
            value: t.variable({ name: PATCHER_FIELD })
          }),
          t.objectField({
            name: 'patch',
            value: t.objectValue({
              fields: defn.fields.nodes
                .filter((field) => field.name !== PATCHER_FIELD)
                .map((field) =>
                  t.objectField({
                    name: fieldMap[field.name].fieldName,
                    value: t.variable({
                      name: fieldMap[field.name].fieldName
                    })
                  })
                )
            })
          })
        ]
      })
    })
  ];

  const selections = defn.fields.nodes.map((field) =>
    t.field({ name: fieldMap[field.name].fieldName })
  );
  const opSel = [
    t.field({
      name: operationName,
      args: selectArgs,
      selectionSet: t.selectionSet({
        selections: [
          t.field({
            name: modelName,
            selectionSet: t.selectionSet({ selections })
          })
        ]
      })
    })
  ];

  const ast = t.document({
    definitions: [
      t.operationDefinition({
        operation: 'mutation',
        name: mutationName,
        variableDefinitions,
        selectionSet: t.selectionSet({ selections: opSel })
      })
    ]
  });

  return { name: mutationName, ast };
};

const GETTER_FIELD = 'id';
export const getOne = (defn) => {
  const operationName = inflection.camelize(
    [plz.singular(defn.name)].join('_'),
    true
  );
  const queryName = inflection.camelize(
    ['get', plz.singular(defn.name), 'query'].join('_'),
    true
  );
  const modelName = inflection.camelize(
    [plz.singular(defn.name)].join('_'),
    true
  );

  const fieldMap = defn.fields.nodes.reduce((m, field) => {
    const fieldName = inflection.camelize(plz.singular(field.name), true);
    const fieldType = getType(field.type);
    const isArray = !!field.isArray;
    const isRequired =
      field.isRequired &&
      (typeof field.defaultValue === 'undefined' ||
        field.defaultValue === null);
    m[field.name] = {
      fieldName,
      fieldType,
      isRequired,
      isArray
    };
    return m;
  }, {});

  const variableDefinitions = defn.fields.nodes
    .filter((field) => field.name === GETTER_FIELD)
    .map((field) => {
      const { fieldName, fieldType, isRequired, isArray } = fieldMap[
        field.name
      ];
      let type = t.namedType({ type: fieldType });
      if (isArray) type = t.listType({ type });
      // later we can use primary key info...
      if (field.name === GETTER_FIELD) type = t.nonNullType({ type });
      return t.variableDefinition({
        variable: t.variable({ name: fieldName }),
        type
      });
    });

  const selectArgs = [
    t.argument({
      name: GETTER_FIELD,
      value: t.variable({ name: GETTER_FIELD })
    })
  ];

  const selections = defn.fields.nodes.map((field) =>
    t.field({ name: fieldMap[field.name].fieldName })
  );
  const opSel = [
    t.field({
      name: operationName,
      args: selectArgs,
      selectionSet: t.selectionSet({ selections })
    })
  ];

  const ast = t.document({
    definitions: [
      t.operationDefinition({
        operation: 'query',
        name: queryName,
        variableDefinitions,
        selectionSet: t.selectionSet({ selections: opSel })
      })
    ]
  });

  return { name: queryName, ast };
};

export const getMany = (defn) => {
  const operationName = inflection.camelize(
    [plz.plural(defn.name)].join('_'),
    true
  );
  const queryName = inflection.camelize(
    ['get', plz.plural(defn.name), 'query'].join('_'),
    true
  );

  const fieldMap = defn.fields.nodes.reduce((m, field) => {
    const fieldName = inflection.camelize(plz.singular(field.name), true);
    const fieldType = getType(field.type);
    const isArray = !!field.isArray;
    const isRequired =
      field.isRequired &&
      (typeof field.defaultValue === 'undefined' ||
        field.defaultValue === null);
    m[field.name] = {
      fieldName,
      fieldType,
      isRequired,
      isArray
    };
    return m;
  }, {});

  const selections = defn.fields.nodes.map((field) =>
    t.field({ name: fieldMap[field.name].fieldName })
  );
  const opSel = [
    t.field({
      name: operationName,
      selectionSet: t.objectValue({
        fields: [
          t.field({
            name: 'totalCount'
          }),
          t.objectField({
            name: 'nodes',
            value: t.selectionSet({ selections })
          })
        ]
      })
    })
  ];

  const ast = t.document({
    definitions: [
      t.operationDefinition({
        operation: 'query',
        name: queryName,
        selectionSet: t.selectionSet({ selections: opSel })
      })
    ]
  });

  return { name: queryName, ast };
};

export const getManyOwned = (defn, ownedField) => {
  const operationName = inflection.camelize(
    [plz.plural(defn.name)].join('_'),
    true
  );
  const queryName = inflection.camelize(
    [
      'get',
      plz.plural(defn.name),
      'owned',
      'by',
      plz.singular(ownedField),
      'query'
    ].join('_'),
    true
  );

  const fieldMap = defn.fields.nodes.reduce((m, field) => {
    const fieldName = inflection.camelize(plz.singular(field.name), true);
    const fieldType = getType(field.type);
    const isArray = !!field.isArray;
    const isRequired =
      field.isRequired &&
      (typeof field.defaultValue === 'undefined' ||
        field.defaultValue === null);
    m[field.name] = {
      fieldName,
      fieldType,
      isRequired,
      isArray
    };
    return m;
  }, {});

  const variableDefinitions = defn.fields.nodes
    .filter((field) => field.name === ownedField)
    .map((field) => {
      const { fieldName, fieldType, isRequired, isArray } = fieldMap[
        field.name
      ];
      let type = t.namedType({ type: fieldType });
      if (isArray) type = t.listType({ type });
      // required
      type = t.nonNullType({ type });
      return t.variableDefinition({
        variable: t.variable({ name: fieldName }),
        type
      });
    });

  const ownedFieldName = inflection.camelize(ownedField, true);

  const condition = [
    t.argument({
      name: 'condition',
      value: t.objectValue({
        fields: [
          t.objectField({
            name: ownedFieldName,
            value: t.variable({ name: ownedFieldName })
          })
        ]
      })
    })
  ];

  const selections = defn.fields.nodes.map((field) =>
    t.field({ name: fieldMap[field.name].fieldName })
  );
  const opSel = [
    t.field({
      name: operationName,
      args: condition,
      selectionSet: t.objectValue({
        fields: [
          t.field({
            name: 'totalCount'
          }),
          t.objectField({
            name: 'nodes',
            value: t.selectionSet({ selections })
          })
        ]
      })
    })
  ];

  const ast = t.document({
    definitions: [
      t.operationDefinition({
        operation: 'query',
        name: queryName,
        variableDefinitions,
        selectionSet: t.selectionSet({ selections: opSel })
      })
    ]
  });

  return { name: queryName, ast };
};

const DELETER_FIELD = 'id';
export const deleteOne = (defn) => {
  const operationName = inflection.camelize(
    ['delete', plz.singular(defn.name)].join('_'),
    true
  );
  const mutationName = inflection.camelize(
    ['delete', plz.singular(defn.name), 'mutation'].join('_'),
    true
  );
  const modelName = inflection.camelize(
    [plz.singular(defn.name)].join('_'),
    true
  );

  const fieldMap = defn.fields.nodes.reduce((m, field) => {
    const fieldName = inflection.camelize(plz.singular(field.name), true);
    const fieldType = getType(field.type);
    const isArray = !!field.isArray;
    const isRequired =
      field.isRequired &&
      (typeof field.defaultValue === 'undefined' ||
        field.defaultValue === null);
    m[field.name] = {
      fieldName,
      fieldType,
      isRequired,
      isArray
    };
    return m;
  }, {});

  const variableDefinitions = defn.fields.nodes
    .filter((field) => field.name === DELETER_FIELD)
    .map((field) => {
      const { fieldName, fieldType, isRequired, isArray } = fieldMap[
        field.name
      ];
      let type = t.namedType({ type: fieldType });
      if (isArray) type = t.listType({ type });

      // required
      type = t.nonNullType({ type });
      return t.variableDefinition({
        variable: t.variable({ name: fieldName }),
        type
      });
    });

  const selectArgs = [
    t.argument({
      name: 'input',
      value: t.objectValue({
        fields: [
          t.objectField({
            name: PATCHER_FIELD,
            value: t.variable({ name: PATCHER_FIELD })
          })
        ]
      })
    })
  ];

  const selections = defn.fields.nodes.map((field) =>
    t.field({ name: fieldMap[field.name].fieldName })
  );
  const opSel = [
    t.field({
      name: operationName,
      args: selectArgs,
      selectionSet: t.selectionSet({
        selections: [
          t.field({
            name: modelName,
            selectionSet: t.selectionSet({ selections })
          })
        ]
      })
    })
  ];

  const ast = t.document({
    definitions: [
      t.operationDefinition({
        operation: 'mutation',
        name: mutationName,
        variableDefinitions,
        selectionSet: t.selectionSet({ selections: opSel })
      })
    ]
  });

  return { name: mutationName, ast };
};
