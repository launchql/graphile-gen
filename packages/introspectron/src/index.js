export * from './introspect';
export * from './process';
export * from './query';
