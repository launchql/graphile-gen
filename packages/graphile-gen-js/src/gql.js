import { gql } from 'graphile-gen';
import { generateJS } from './utils';

export const generate = (introspectron) => {
  const val = gql.generate(introspectron);
  return Object.keys(val).reduce((m, key) => {
    m[key] = generateJS({ name: key, ast: val[key].ast });
    return m;
  }, {});
};
