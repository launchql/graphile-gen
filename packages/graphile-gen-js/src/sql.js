import { pg } from 'graphile-gen';
import { generateJS } from './utils';

export const crudify = (introspectron) => {
  const val = pg.crudify(introspectron);
  return Object.keys(val).reduce((m, key) => {
    m[key] = generateJS({ name: key, ast: val[key] });
    return m;
  }, {});
};

export const owned = (introspectron) => {
  const val = pg.owned(introspectron);
  return Object.keys(val).reduce((m, key) => {
    m[key] = generateJS(val[key]);
    return m;
  }, {});
};

export const updateOne = (defn) => {
  const val = pg.updateOne(defn);
  return generateJS(val);
};

export const createOne = (defn) => {
  const val = pg.createOne(defn);
  return generateJS(val);
};

export const deleteOne = (defn) => {
  const val = pg.deleteOne(defn);
  return generateJS(val);
};

export const getOne = (defn) => {
  const val = pg.getOne(defn);
  return generateJS(val);
};

export const getMany = (defn) => {
  const val = pg.getMany(defn);
  return generateJS(val);
};

export const getManyOwned = (defn, ownedKey) => {
  const val = pg.getManyOwned(defn, ownedKey);
  return generateJS(val);
};
