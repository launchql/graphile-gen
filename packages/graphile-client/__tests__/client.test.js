import mutations from '../__fixtures__/api/mutations.json';
import queries from '../__fixtures__/api/queries.json';
import { GraphileClient } from '../src';

it('getMany', () => {
  const client = new GraphileClient({ ...queries, ...mutations });
  const result = client
    .query('Action')
    .select({
      id: true,
      name: true,
      photo: true,
      title: true,
      actionResults: {
        select: {
          id: true
        },
        variables: {
          first: 10,
          last: 10,
          before: null
        }
      }
    })
    .getMany()
    .print();
  expect(client._hash).toMatchSnapshot();
  expect(client._queryName).toMatchSnapshot();
});

it('getMany edges', () => {
  const client = new GraphileClient({ ...queries, ...mutations });
  const result = client
    .query('Action')
    .select({
      id: true,
      name: true,
      photo: true,
      title: true,
      actionResults: {
        select: {
          id: true
        },
        variables: {
          first: 10,
          last: 10,
          before: null
        }
      }
    })
    .edges(true)
    .getMany()
    .print();
  expect(client._hash).toMatchSnapshot();
  expect(client._queryName).toMatchSnapshot();
});

it('getOne', () => {
  const client = new GraphileClient({ ...queries, ...mutations });
  const result = client
    .query('Action')
    .select({
      id: true,
      name: true,
      photo: true,
      title: true,
      actionResults: {
        select: {
          id: true
        },
        variables: {
          first: 10,
          last: 10,
          before: null
        }
      }
    })
    .getOne()
    .print();
  expect(client._hash).toMatchSnapshot();
  expect(client._queryName).toMatchSnapshot();
});

it('getAll', () => {
  const client = new GraphileClient({ ...queries, ...mutations });
  const result = client
    .query('Action')
    .select({
      id: true,
      name: true,
      photo: true,
      title: true,
      actionResults: {
        select: {
          id: true
        },
        variables: {
          first: 10,
          last: 10,
          before: null
        }
      }
    })
    .all()
    .print();
  expect(client._hash).toMatchSnapshot();
  expect(client._queryName).toMatchSnapshot();
});

it('create', () => {
  const client = new GraphileClient({ ...queries, ...mutations });
  const result = client
    .query('Action')
    .select({
      id: true,
      name: true,
      photo: true,
      title: true
    })
    .create()
    .print();
  expect(client._hash).toMatchSnapshot();
  expect(client._queryName).toMatchSnapshot();
});
it('delete', () => {
  const client = new GraphileClient({ ...queries, ...mutations });
  const result = client
    .query('Action')
    .select({
      id: true,
      name: true,
      photo: true,
      title: true
    })
    .delete()
    .print();
  expect(client._hash).toMatchSnapshot();
  expect(client._queryName).toMatchSnapshot();
});
it('update', () => {
  const client = new GraphileClient({ ...queries, ...mutations });
  const result = client
    .query('Action')
    .select({
      id: true,
      name: true,
      photo: true,
      title: true
    })
    .update()
    .print();
  expect(client._hash).toMatchSnapshot();
  expect(client._queryName).toMatchSnapshot();
});
