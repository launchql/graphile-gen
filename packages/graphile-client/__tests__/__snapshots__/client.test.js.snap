// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAll 1`] = `
"query getActionsQueryAll {
  actions {totalCount, nodes {
    id
    name
    photo
    title
    actionResults
  }}
}
"
`;

exports[`getAll 2`] = `"getActionsQueryAll"`;

exports[`getMany 1`] = `
"query getActionsQuery($first: Int, $last: Int, $after: Cursor, $before: Cursor, $offset: Int, $condition: ActionCondition, $filter: ActionFilter, $orderBy: [ActionsOrderBy!]) {
  actions(first: $first, last: $last, offset: $offset, after: $after, before: $before, condition: $condition, filter: $filter, orderBy: $orderBy) {totalCount, pageInfo {
    hasNextPage
    hasPreviousPage
    endCursor
    startCursor
  }, nodes {
    id
    name
    photo
    title
    actionResults
  }}
}
"
`;

exports[`getMany 2`] = `"getActionsQuery"`;

exports[`getMany edges 1`] = `
"query getActionsQuery($first: Int, $last: Int, $after: Cursor, $before: Cursor, $offset: Int, $condition: ActionCondition, $filter: ActionFilter, $orderBy: [ActionsOrderBy!]) {
  actions(first: $first, last: $last, offset: $offset, after: $after, before: $before, condition: $condition, filter: $filter, orderBy: $orderBy) {totalCount, pageInfo {
    hasNextPage
    hasPreviousPage
    endCursor
    startCursor
  }, edges {
    cursor
    node {
      id
      name
      photo
      title
      actionResults
    }
  }}
}
"
`;

exports[`getMany edges 2`] = `"getActionsQuery"`;

exports[`getOne 1`] = `
"query getActionQuery($id: UUID!) {
  action(id: $id) {
    id
    name
    photo
    title
    actionResults
  }
}
"
`;

exports[`getOne 2`] = `"getActionQuery"`;
